pkg/kv/kvserver/mvcc_gc_queue.go

```diff
@@ -722,9 +782,78 @@ func updateStoreMetricsWithGCInfo(metrics *StoreMetrics, info gc.Info) {
 	metrics.GCFailedClearRange.Inc(int64(info.ClearRangeKeyFailures))
 }
 
+func (mgcq *mvccGCQueue) postProcessScheduled(
+	ctx context.Context, processedReplica replicaInQueue, priority float64,
+) {
+	if priority < deleteRangePriority {
+		mgcq.lastRangeWasHighPriority = false
+		return
+	}
+
+	if !mgcq.lastRangeWasHighPriority {
+		// We are most likely processing first range that has a GC hint notifying
+		// that multiple range deletions happen.
+		mgcq.scanReplicasForHiPriGCHints(ctx, processedReplica.GetRangeID())
+		// Set flag indicating that we are already collecting high priority to avoid
+		// rescanning and re-enqueueing ranges multiple times.
+		mgcq.lastRangeWasHighPriority = true
+	}
+}
+
+func (mgcq *mvccGCQueue) scanReplicasForHiPriGCHints(
+	ctx context.Context, triggerRange roachpb.RangeID,
+) {
+	// We run scanner for a limited time to ensure GC queue can make progress if
+	// something goes wrong with leaseholders. Since visitor is shuffling replicas
+	// prior to iteration we should eventually process all even if first attempt
+	// fails.
+	var foundRanges int
+	err := contextutil.RunWithTimeout(ctx, "gc-check-hinted-hi-pri-replicas", gcHintScannerTimeout, func(ctx context.Context) error {
+		v := newStoreReplicaVisitor(mgcq.store)
+		v.Visit(func(replica *Replica) bool {
+			if replica.GetRangeID() == triggerRange {
+				return true
+			}
+			gCHint := replica.GetGCHint()
+			if !gCHint.LatestRangeDeleteTimestamp.IsEmpty() {
+				desc, spanConfig := replica.DescAndSpanConfig()
+				gcThreshold := replica.store.Clock().Now().Add(-int64(spanConfig.GCPolicy.TTLSeconds)*1e9,
```

Comment:
Can we try to avoid the `Now()` calls, and just take a clock reading at the start of the loop or something? These take out exclusive locks on the clock and such, serializing with all other clock readers, so it seems worthwhile to avoid calling it in a tight loop.